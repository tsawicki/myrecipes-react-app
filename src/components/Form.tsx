import React, { useState, FormEvent } from "react";
import { Recipe } from "../models/Recipe.model";

interface OwnProps {
  recipe: Recipe;
  onSubmit: (recipe: Recipe) => void;
}

const Form: React.FunctionComponent<OwnProps> = ({
  recipe,
  onSubmit
}: OwnProps): JSX.Element => {
  const noErrors = {
    name: "",
    ingredients: ""
  };

  const [input, setInput] = useState(recipe);
  const [errors, setErrors] = useState(noErrors);

  type input = HTMLInputElement | HTMLTextAreaElement;

  const handleInputChange = (event: React.FormEvent<input>) =>
    setInput({
      ...input,
      [event.currentTarget.name]: event.currentTarget.value
    });

  const validateForm = () => {
    let isFormValid = true;
    let newErrors = noErrors;

    if (input.name === "") {
      newErrors = {
        ...newErrors,
        name: "Name can't be blank"
      };
      isFormValid = false;
    }

    if (input.ingredients === "") {
      newErrors = {
        ...newErrors,
        ingredients: "Add at least one ingredient"
      };
      isFormValid = false;
    }
    setErrors(newErrors);
    return isFormValid;
  };

  const handleSubmit = (event: FormEvent) => {
    event.preventDefault();
    if (validateForm()) {
      onSubmit(input);
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <label>Recipe Name*:</label>
        <input
          type="text"
          name="name"
          className={errors.name !== "" ? "has-error" : ""}
          onChange={handleInputChange}
          value={input.name}
        />
        {errors.name !== "" && (
          <small className="error-msg">{errors.name}</small>
        )}
      </div>
      <div>
        <label>Ingredients*:</label>
        <input
          type="text"
          name="ingredients"
          className={errors.ingredients !== "" ? "has-error" : ""}
          onChange={handleInputChange}
          value={input.ingredients}
        />
        {errors.ingredients !== "" && (
          <small className="error-msg">{errors.ingredients}</small>
        )}
      </div>
      <div>
        <label>Directions:</label>
        <textarea
          name="directions"
          rows={5}
          onChange={handleInputChange}
          value={input.directions}
        />
      </div>
      <input
        type="submit"
        className="btn"
        value="Save Recipe"
        data-testid="submit-form-btn"
      />
    </form>
  );
};

export default Form;
