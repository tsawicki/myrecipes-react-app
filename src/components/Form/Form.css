label {
  font-size: 10px;
  text-transform: uppercase;
  font-size: 0.55em;
  letter-spacing: 0.3em;
  color: #999;
  display: block;
}

input,
textarea {
  border-radius: 10px;
  outline: none;
  border: 1px solid #dedede;
  margin-top: 10px;
  margin-bottom: 20px;
  width: 100%;
  padding: 10px 15px;
  font-weight: 300;
}

input:focus,
textarea:focus {
  box-shadow: 0px 0px 10px 0px rgba(0, 0, 0, 0.05);
}

.has-error {
  border: 1px solid red;
}

.error-msg {
  color: red;
  font-size: 11px;
  position: relative;
  bottom: 20px;
}
