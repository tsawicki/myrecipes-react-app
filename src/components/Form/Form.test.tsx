import React from "react";
import { render } from "@testing-library/react";
import Form from "./Form";

test("Form component", () => {
  describe("matches the snapshot", () => {
    const wrapper = render(
      <Form
        onSubmit={() => {}}
        recipe={{
          id: "1234",
          name: "test recipe",
          ingredients: "test ingredients",
          directions: ""
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});

test("Form component validates the name field", () => {
  const onSubmit = jest.fn();
  const wrapper = render(
    <Form
      onSubmit={onSubmit}
      recipe={{
        id: "1234",
        name: "",
        ingredients: "test ingredients",
        directions: ""
      }}
    />
  );
  wrapper.getByTestId("submit-form-btn").click();
  expect(onSubmit).not.toBeCalled();
  expect(wrapper.getByText(/Name can't be blank/)).toBeInTheDocument();
});

test("Form component validates the ingredients field", () => {
  const onSubmit = jest.fn();
  const wrapper = render(
    <Form
      onSubmit={onSubmit}
      recipe={{
        id: "1234",
        name: "test recipe",
        ingredients: "",
        directions: ""
      }}
    />
  );
  wrapper.getByTestId("submit-form-btn").click();
  expect(onSubmit).not.toBeCalled();
  expect(wrapper.getByText(/Add at least one ingredient/)).toBeInTheDocument();
});
