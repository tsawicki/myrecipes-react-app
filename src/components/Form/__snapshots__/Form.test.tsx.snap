// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div>
          <label>
            Recipe Name*:
          </label>
          <input
            class=""
            name="name"
            type="text"
            value="test recipe"
          />
        </div>
        <div>
          <label>
            Ingredients*:
          </label>
          <input
            class=""
            name="ingredients"
            type="text"
            value="test ingredients"
          />
        </div>
        <div>
          <label>
            Directions:
          </label>
          <textarea
            name="directions"
            rows="5"
          />
        </div>
        <input
          class="btn"
          data-testid="submit-form-btn"
          type="submit"
          value="Save Recipe"
        />
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div>
        <label>
          Recipe Name*:
        </label>
        <input
          class=""
          name="name"
          type="text"
          value="test recipe"
        />
      </div>
      <div>
        <label>
          Ingredients*:
        </label>
        <input
          class=""
          name="ingredients"
          type="text"
          value="test ingredients"
        />
      </div>
      <div>
        <label>
          Directions:
        </label>
        <textarea
          name="directions"
          rows="5"
        />
      </div>
      <input
        class="btn"
        data-testid="submit-form-btn"
        type="submit"
        value="Save Recipe"
      />
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
